src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java

```diff
@@ -0,0 +1,298 @@
+    private static final String CTP_PRODUCT_TYPE_CREATE_FAILED = "Failed to create product type of key '%s'.";
+    private static final String PRODUCT_TYPE_DRAFT_HAS_NO_KEY = "Failed to process product type draft without key.";
+    private static final String PRODUCT_TYPE_DRAFT_IS_NULL = "Failed to process null product type draft.";
+
+    private final ProductTypeService productTypeService;
+
+    /**
+     * Takes a {@link ProductTypeSyncOptions} instance to instantiate a new {@link ProductTypeSync} instance that
+     * could be used to sync product type drafts with the given product types in the CTP project specified in the
+     * injected {@link ProductTypeSyncOptions} instance.
+     *
+     * @param productTypeSyncOptions the container of all the options of the sync process including the CTP project
+     *                               client and/ormconfiguration and other sync-specific options.
+     */
+    public ProductTypeSync(@Nonnull final ProductTypeSyncOptions productTypeSyncOptions) {
+        super(new ProductTypeSyncStatistics(), productTypeSyncOptions);
+        this.productTypeService = new ProductTypeServiceImpl(productTypeSyncOptions);
+    }
+
+    /**
+     * Takes a {@link ProductTypeSyncOptions} instance to instantiate a new {@link ProductTypeSync} instance that
+     * could be used to sync product type drafts with the given product types in the CTP project specified in the
+     * injected {@link ProductTypeSyncOptions} instance.
+     *
+     * @param productTypeSyncOptions the container of all the options of the sync process including the CTP project
+     *                               client and/ormconfiguration and other sync-specific options.
+     * @param productTypeService     the product type service to make product type requests.
+     *
+     *
+     *
+     */
+    public ProductTypeSync(@Nonnull final ProductTypeSyncOptions productTypeSyncOptions,
+                           @Nonnull final ProductTypeService productTypeService) {
+        super(new ProductTypeSyncStatistics(), productTypeSyncOptions);
+        this.productTypeService = productTypeService;
+    }
+
+    /**
+     * Iterates through the whole {@code productTypeDrafts} list and accumulates its valid drafts to batches.
+     * Every batch is then processed by {@link ProductTypeSync#processBatch(List)}.
+     *
+     * <p><strong>Inherited doc:</strong>
+     * {@inheritDoc}
+     *
+     * @param productTypeDrafts {@link List} of {@link ProductTypeDraft}'s that would be synced into CTP project.
+     * @return {@link CompletionStage} with {@link ProductTypeSyncStatistics} holding statistics of all sync
+     *         processes performed by this sync instance.
+     */
+    @Override
+    protected CompletionStage<ProductTypeSyncStatistics> process(
+            @Nonnull final List<ProductTypeDraft> productTypeDrafts) {
+
+        final List<List<ProductTypeDraft>> batches = batchElements(productTypeDrafts, syncOptions.getBatchSize());
```

Comment:
The invalid drafts count as a processed items in the statistics so they have to be "processed" (they are not actually processed, but the process method will consider them as a processed items)

Reply:
What I meant is this:
```
final List<ProductTypeDraft> validProductTypeDrafts = productTypeDrafts.stream()
                                                                   .filter(this::validateDraft)
                                                                   .collect(toList());

        final List<List<ProductTypeDraft>> batches = batchElements(validProductTypeDrafts, syncOptions.getBatchSize());
```

and remove `validateDraft` part from `processBatch()` method.

Reply:
mmm... but if I validate there, then the counter for "processed" items is not increased in `processBatch` because the invalid drafts are removed before and `processBatch` is responsible for increasing the counter of processed items in the statistics. 

In the other synchronization of entities, the invalid drafts are considered as processed. Furthermore, in the `processBatch` the invalid drafts are filtered before processing, so it seems that there is no performance advantage of moving the filtering to the process.

Maybe I'm wrong, what do you think @heshamMassoud ?

Reply:
No, you are arguments are absolutely correct @jortizsao.